### ISSUE TITLE
Batch sizes for 24gb gpu

### ISSUE THREAD

**Issue by Ph0rk0z** [mit-han-lab/deepcompressor#66]
Anyone tried to quantize on 24gb gpu or multiple 24g gpus? The batch sizes in the default configs are obviously set up for cloud host big vram cards but in theory it should be possible, right?

Otherwise kind of hard if you have a flux (or other model) with finetune and nobody has paid and rented servers to compress it.Yes, it will probably take longer but that's better than nothing.

Might become more relevant as models like this become more common: https://huggingface.co/lodestones/Chroma along with the flux de-distills. Projects such as exllama and GPTQ/AWQ itself (i think) can quant a 70b in 24gb over time and make the format much more accessible.

**Comment by josephrocca** [comment]
I have a related question: I'm also wondering if `quant.calib.batch_size=1` is valid?

I.e. are higher batch sizes just to speed up the process, or are they relevant to the actual accuracy of the final quantized model?

Concretely, I'm wondering if this is fine:
```
sed -i 's/    batch_size: 16/    batch_size: 1/' /workspace/deepcompressor/examples/diffusion/configs/model/flux.1-schnell.yaml
```

**Comment by josephrocca** [comment]
> I'm also wondering if `quant.calib.batch_size=1` is valid?

I've just asked Gemini 2.5 Pro about this and it says it's fine to use `quant.calib.batch_size=1`. Below is the answer, for whatever it is worth. Obviously can't trust it as the final answer at all, but it seems to make sense. The most relevant parts are the last point in section 3, and all of section 4.

---

1.  **Configuration:**
    *   The parameter is defined in `examples/diffusion/configs/__default__.yaml` under `quant.calib`.
    *   The `DiffusionCalibCacheLoaderConfig` dataclass in `deepcompressor/app/diffusion/dataset/calib.py` holds this configuration. It inherits from `BaseDataLoaderConfig`.

2.  **Loading the Calibration Data:**
    *   The primary use of `quant.calib.batch_size` is within the `DiffusionCalibCacheLoader` class (also in `deepcompressor/app/diffusion/dataset/calib.py`).
    *   Specifically, in the `iter_samples` method, it builds a standard `torch.utils.data.DataLoader`:
        ```python
        dataloader = self.dataset.build_loader(
            batch_size=self.batch_size, # <-- Here it's used
            shuffle=False,
            drop_last=True,
            num_workers=self.config.num_workers
        )
        for data in dataloader:
            yield ModuleForwardInput(args=data["input_args"], kwargs=data["input_kwargs"])
        ```
    *   This means `quant.calib.batch_size` controls how many calibration samples are loaded from the dataset and passed through the model *at one time* during the calibration *data collection phase*.

3.  **Collecting Activation Statistics:**
    *   The `iter_layer_activations` method (and its underlying `_iter_layer_activations` from the base class `BaseCalibCacheLoader`) iterates through the samples yielded by `iter_samples`.
    *   The `action` argument, typically an instance of `ConcatCacheAction` (defined in `deepcompressor/dataset/action.py`), is used to process the activations generated during these forward passes.
    *   `ConcatCacheAction.apply` takes the tensors from the current batch and concatenates them to previously collected tensors along the 0th dimension (the sample/batch dimension).
        ```python
        # Simplified logic from ConcatCacheAction.apply
        c.data[0][num_prev_cached : c.num_cached].copy_(x)
        ```
    *   This means that regardless of the `batch_size` used for loading, the `ConcatCacheAction` accumulates all processed samples into a single large tensor (or set of tensors) representing the *entire* calibration dataset requested (`quant.calib.num_samples`).

4.  **Using the Calibration Data:**
    *   The collected (concatenated) activation tensors stored in the `layer_cache` are then used by various calibration steps:
        *   **Smoothing (`smooth_diffusion`):** Calculates statistics (like `AbsMax` or `RootMeanSquare`) over the *entire* collected activation cache for a given layer/module to determine smoothing scales.
        *   **Low-Rank Branch Calibration (`calibrate_diffusion_block_low_rank_branch`):** Performs operations like SVD on error matrices derived from the *entire* collected activation cache.
        *   **Weight Range Calibration (`update_diffusion_block_weight_quantizer_state_dict` when `config.wgts.needs_calib_data` is true):** Uses the *entire* collected input activation cache to calibrate weight ranges, often involving error minimization or statistical analysis across all samples.
        *   **Activation Range Calibration (`quantize_diffusion_activations` when `config.ipts/opts.needs_calib_data` is true):** Calculates activation ranges (min/max) based on the *entire* collected activation cache.
        *   **GPTQ (`quantize_diffusion_block_weights` when `config.wgts.enabled_gptq` is true):** Computes the Hessian based on the *entire* collected input activation cache and updates weights iteratively.

5.  **Conclusion:**
    *   The `quant.calib.batch_size` only affects how the calibration data is loaded and passed through the model during the *collection* phase.
    *   The actual calibration algorithms (smoothing, SVD, range finding, GPTQ) operate on the *aggregated* data collected from *all* calibration samples (typically concatenated by `ConcatCacheAction`).
    *   The final statistics and derived quantization parameters (smoothing scales, SVD matrices, quantization ranges, GPTQ weight updates) depend on the complete set of calibration samples (`quant.calib.num_samples`), not on the `batch_size` used to load them.

**Comment by Ph0rk0z** [comment]
I don't think batch size is related to the quality. You're still going through the whole dataset just slower. That guy's 6000 hour thing is insane, gotta be bleeding over into sysram or something. Does more than batch 1 not fit?

**Comment by josephrocca** [comment]
You can disable the eval like I mentioned [here](https://github.com/mit-han-lab/deepcompressor/issues/70#issuecomment-2787905770) and then it's like 12 hours or so on a H100 at batch size 1. So if it fits in 24GB VRAM/RAM, I think *at most* it'd take a few days on a 3090 at batch size 1.

**Comment by rwfsmith** [comment]
I also decided to try loading the model with fp8 layerwise casting and text encoder offloading to reduce memory usage on my 5090, seems to be working but I don't know how much that'll affect the quality. 

The eval step still just barely fits in the 32GB VRAM (around 30GB VRAM used right now), though, so might still be too much for a 3090. Also reduced sampling steps to 20, because I'm impatient. That did reduce the time per sample from ~2 hours to around 15 seconds, though, and I'm limiting it to 100 samples... because impatient. Also, just testing it out, anyway, so I don't want it to take forever just to see if it works.

**Comment by rwfsmith** [comment]
well, that didn't help much. Was able to get through the eval and calib steps just fine, but the smooth and quant step still uses way too much vram and I'm not sure what to do about that. I'll try getting it running on my Strix Halo APU which will be a bit slower but has up to around 110GB of VRAM.

**Comment by Ph0rk0z** [comment]
I think they need to implement stuff like FSDP for the quant process. Otherwise we are stuck with renting big GPUs.

**Comment by rwfsmith** [comment]
I've gotten it running in 32GB of VRAM (but still uses a lot of RAM) by using group offloading for the text encoders and reducing batch sizes quite a bit. Smoothing is still very slow and will probably take a couple days to finish. I'm tempted to try skipping it to see how bad it is without it :). Could probably reduce the batch sizes further to get it into 24gb. I think diffusers 0.33.0 is the minimum needed for the group offloading? 


flux.1-dev.yaml:

```
pipeline:
  name: flux.1-dev
  dtype: torch.bfloat16
eval:
  num_steps: 50
  guidance_scale: 3.5
  protocol: fmeuler{num_steps}-g{guidance_scale}
quant:
  calib:
    batch_size: 4 
  wgts:
    calib_range:
      element_batch_size: 8  
      sample_batch_size: 2    
      element_size: 256       
      sample_size: 32         
    low_rank:
      sample_batch_size: 1    
      sample_size: 16         
      num_iters: 1            
    skips: 
    - embed
    - resblock_shortcut
    - resblock_time_proj
    - transformer_proj_in
    - transformer_proj_out
    - down_sample
    - up_sample
  ipts:
    calib_range:
      element_batch_size: 8   
      sample_batch_size: 2    
      element_size: 256      
      sample_size: 32        
    skips: 
    - embed
    - resblock_shortcut
    - resblock_time_proj
    - transformer_proj_in
    - transformer_proj_out
    - transformer_norm
    - transformer_add_norm
    - down_sample
    - up_sample
  opts:
    calib_range:
      element_batch_size: 8   
      sample_batch_size: 2   
      element_size: 256       
      sample_size: 32         
  smooth:
    proj:
      element_batch_size: 8   
      sample_batch_size: 1   
      element_size: 256
      sample_size: 32
    attn:
      sample_batch_size: 1    
      sample_size: 32
```

/deepcompressor/app/diffusion/pipeline/config.py


```
        18: 
from diffusers.hooks import apply_group_offloading

365-374:
        #pipeline = pipeline.to(device)
        model = pipeline.unet if hasattr(pipeline, "unet") else pipeline.transformer
        apply_group_offloading(pipeline.text_encoder, onload_device=onload_device, offload_device=offload_device, offload_type="leaf_level", num_blocks_per_group=1)     
        apply_group_offloading(pipeline.text_encoder_2, onload_device=onload_device, offload_device=offload_device, offload_type="leaf_level", num_blocks_per_group=1)     
        replace_fused_linear_with_concat_linear(model)
        replace_up_block_conv_with_concat_conv(model)
        if shift_activations:
            shift_input_activations(model)
        pipeline.to(device)
        return pipeline
```

**Comment by rwfsmith** [comment]
Takes around 3 hours to calibrate and quantize the model without smoothing and results are decent. Around 40 hours estimated to smooth it. I didn't realize I needed to add --float-point to the nunchaku convert step for it to actually work, so that was annoying :). 

I'm trying the smooth with the fast.yaml config added to see how long it takes.

**Comment by rwfsmith** [comment]
If you add "--calib-num-samples 10" to the smoot/quant script it reduces the RAM used by around 50GB, at least. Seems to only reduce VRAM usage a couple gigs, but that might help out. The number of calibration samples doesn't seem to affect performance much if you have enough RAM to store it, both 10 and64 takes about 20 minutes per layer when the smooth proj num_grids is set to 10. Maybe could try to reduce that to 5 to speed it up further, but I'm not sure how that'll affect quality. 

For fun, I set the grids to 2 and it only takes 4 minutes per layer now, we'll see how it looks compared to skipping smoothing entirely.

**Comment by rwfsmith** [comment]
the quantized model seems to have not received great image quality scores? I'm kinda new to all this stuff. I'll try more grids later when I have more time. Would be nice if I had a dedicated machine for this so I could just let it run for a while and not care :) 

{'with_gt': {'clip_iqa': 0.9028482437133789, 'clip_score': 26.55710792541504, 'image_reward': 0.3760984569787979, 'fid': 230.9644747216766}, 'with_orig': {'lpips': 0.24844102561473846, 'psnr': 20.07721519470215, 'ssim': 0.8008022308349609, 'fid': 137.14506255566792}}

**Comment by Ph0rk0z** [comment]
Even 40h isn't that bad if you can quant models yourself without paying. I'm guessing those times go up on 3090 vs 5090.

No smoothing is better than minimal smoothing? And how does it compare to BF16 or just GGUF int8?

**Comment by rwfsmith** [comment]
When using the same seed, smoothed version looks almost identical to the FP16 fine-tune that I started with, down to freckles being in the same spot and skin details being the same. this version takes around 4 seconds to get through 30 steps and uses <11GB of VRAM if you offload the text encoders to CPU, compared to ~13 seconds and >26GB of VRAM usage, 768x1344 resolution. Works pretty well with the turbo LoRA and can generate an image in less than two seconds with good quality.

**Comment by Ph0rk0z** [comment]
But we have to do the smoothing, right? No way to skip it without bad quality?

**Comment by rwfsmith** [comment]
The quality without smoothing is still pretty good, 2 grids of smoothing alone seems to make it match the original model a little bit closer and only adds a couple hours of time to the process. I'll try it with 10 next to see how it goes.  You can skip it with the "--enable-smooth false" parameter.

**Comment by spooknik** [comment]
> The quality without smoothing is still pretty good, 2 grids of smoothing alone seems to make it match the original model a little bit closer and only adds a couple hours of time to the process. I'll try it with 10 next to see how it goes. You can skip it with the "--enable-smooth false" parameter.

And follow up to this? I'm running 10 and it's going to take 60 hours on an A40 (runpod), it seems like the grid search part is CPU bound since it hammers one core while the GPU power is around 175/300 watts.

**Comment by rwfsmith** [comment]
> And follow up to this? I'm running 10 and it's going to take 60 hours on an A40 (runpod), it seems like the grid search part is CPU bound since it hammers one core while the GPU power is around 175/300 watts.

I'm not an expert in this, just a tinkerer that gets bored and moves on quickly. I haven't done anything with it since my last comment. I was doing this all locally on my computer and have never used runpod.

**Comment by Ahmaddimran** [comment]
has anyone here been able to do inference using the quantized model checkpoints, I am encountering error when passing the resulting config in the run to load function. @rwfsmith @josephrocca
